src/net/client/cache.rs

```diff
@@ -0,0 +1,1177 @@
+                    let value =
+                        Arc::new(Value::<C>::new_from_value_and_response(
+                            value.clone(),
+                            Ok(msg),
+                            &self.config,
+                        )?);
+                    self.cache_insert(key.clone(), value.clone()).await;
+                    return Ok(Some(value));
+                }
+            }
+        }
+        Ok(opt_value)
+    }
+
+    /// Insert new entry in the cache.
+    ///
+    /// Do not insert if the validity is zero.
+    /// Make sure to clear the AA flag.
+    async fn cache_insert(&self, key: Key, value: Arc<Value<C>>) {
+        if value.valid_for.is_zero() {
+            // Do not insert cache value that are valid for zero duration.
+            return;
+        }
+        let value = match prepare_for_insert(value.clone(), &self.config) {
+            Ok(value) => value,
+            Err(e) => {
+                // Create a new value based on this error
+                Arc::new(
+                    Value::<C>::new_from_value_and_response(
+                        value,
+                        Err(e),
+                        &self.config,
+                    )
+                    .expect("value from error does not fail"),
+                )
+            }
+        };
+        self.cache.insert(key, value).await
+    }
+}
+
+impl<CR, Upstream, C> Debug for Request<CR, Upstream, C>
+where
+    CR: Send + Sync,
+    Upstream: Send + Sync,
+    C: Clock + Send + Sync,
+{
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), core::fmt::Error> {
+        f.debug_struct("Request")
+            .field("fut", &format_args!("_"))
+            .finish()
+    }
+}
+
+impl<CR, Upstream, C> GetResponse for Request<CR, Upstream, C>
+where
+    CR: Clone + ComposeRequest + Debug + Sync,
+    Upstream: SendRequest<CR> + Send + Sync + 'static,
+    C: Clock + Debug + Send + Sync + 'static,
+{
+    fn get_response(
+        &mut self,
+    ) -> Pin<
+        Box<
+            dyn Future<Output = Result<Message<Bytes>, Error>>
+                + Send
+                + Sync
+                + '_,
+        >,
+    > {
+        Box::pin(self.get_response_impl())
+    }
+}
+
+//------------ RequestState ---------------------------------------------------
+/// States of the state machine in get_response_impl
+enum RequestState {
+    /// Initial state, perform a cache lookup.
+    Init,
+
+    /// Wait for a response and insert the response in the cache.
+    GetResponse(Key, Box<dyn GetResponse + Send + Sync>),
+
+    /// Wait for a response but do not insert the response in the cache.
+    GetResponseNoCache(Box<dyn GetResponse + Send + Sync>),
+}
+
+//------------ Key ------------------------------------------------------------
+
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+/// The key for cache entries.
+struct Key {
+    /// DNS name in the request.
+    qname: Dname<Vec<u8>>,
```

Comment:
Given that you are cloning the key quite a lot, it might be beneficial to use a `Dname<Bytes>` here since it is a lot cheaper to clone?

Reply:
Bytes does not implement AsRef<[u8]> which is needed for Debug.

Reply:
It certainly does: https://docs.rs/bytes/latest/bytes/struct.Bytes.html#impl-AsRef%3C%5Bu8%5D%3E-for-Bytes

Reply:
I must have done something wrong. Now I get a new error: 

760 |         qname.make_canonical();
    |               ^^^^^^^^^^^^^^ the trait `AsMut<[u8]>` is not implemented for `bytes::Bytes`

I can start out with a Vec<u8> but then how do I convert to Bytes?

Reply:
Just call `.into()` on it.

Reply:
763 |     let qname: Dname<Bytes> = qname.into();
    |                                     ^^^^ the trait `From<name::dname::Dname<Vec<u8>>>` is not implemented for `name::dname::Dname<bytes::Bytes>`

Reply:
Ah, sorry. The `OctetsInto` is the version of `Into` that converts the underlying octets, i.e., `qname.octets_into()` with `use octseq::octets::OctetsInto`.

Reply:
Yes, that worked!